### PR TITLE
ENH: Add val_set_is_dist flag for small validation sets

### PR DESCRIPTION
**Pull request by kenza-bouzid**
Do not shard data: make sure all processes validate on the entire validation set (reduce will give the same mean)
Log extra validation outputs with a different prefix
Refactor ClassifierParams

### CODE REVIEW

File: hi-ml-cpath/src/health_cpath/configs/classification/BaseMIL.py [hunk 1]
```diff
@@ -72,27 +72,25 @@ class BaseMIL(LightningContainer, EncoderParams, PoolingParams, LossCallbackPara
                                                              "generating outputs.")
     maximise_primary_metric: bool = param.Boolean(True, doc="Whether the primary validation metric should be "
                                                             "maximised (otherwise minimised).")
-    tune_classifier: bool = param.Boolean(
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/configs/classification/BaseMIL.py:75]: For consistency with the rest of the network, encapsulate classifier specific parameters in ClassifierParams class

File: hi-ml-cpath/src/health_cpath/models/deepmil.py [hunk 2]
```diff
@@ -79,35 +79,40 @@ def __init__(self,
...
         self.optimizer_params = optimizer_params
 
         self.save_hyperparameters()
         self.verbose = verbose
         self.outputs_handler = outputs_handler
-        self.pretrained_classifier = pretrained_classifier
 
         # This flag can be switched on before invoking trainer.validate() to enable saving additional time/memory
         # consuming validation outputs via calling self.on_run_extra_validation_epoch()
-        self._run_extra_val_epoch = False
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/models/deepmil.py:94]: renamed to reflect a state

File: hi-ml-cpath/src/health_cpath/models/deepmil.py [hunk 3]
```diff
@@ -363,34 +363,30 @@ def training_step(self, batch: Dict, batch_idx: int) -> BatchResultsType:  # typ
...
                 epoch_results=epoch_results,
                 metrics_dict=self.get_metrics_dict(ModelKey.VAL),  # type: ignore
                 epoch=self.current_epoch,
                 is_global_rank_zero=self.global_rank == 0,
-                run_extra_val_epoch=self._run_extra_val_epoch
+                on_extra_val=self._on_extra_val_epoch,
+                sync_dist=self.should_sync_dist_val(),
             )
 
-            # Reset the top and bottom slides heaps
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/models/deepmil.py:390]: moved this to outputs, it should be handled by it instead of deepmil

File: hi-ml-cpath/src/health_cpath/datamodules/base_module.py [hunk 4]
```diff
@@ -251,13 +261,13 @@ class SlidesDataModule(HistoDataModule[SlidesDataset]):
 
     def __init__(
         self,
-        level: Optional[int] = 1,
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/datamodules/base_module.py:254]: Wrong typing

File: hi-ml-cpath/src/health_cpath/utils/output_utils.py [hunk 5]
```diff
@@ -22,15 +22,17 @@
...
 PREV_VAL_OUTPUTS_SUBDIR = "val_old"
 TEST_OUTPUTS_SUBDIR = "test"
+EXTRA_VAL_OUTPUTS_SUBDIR = "extra_val"
+EXTRA_PREFIX = "extra_"
 
 AML_OUTPUTS_DIR = "outputs"
-AML_LEGACY_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, OUTPUTS_CSV_FILENAME])
-AML_VAL_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, VAL_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME])
-AML_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, TEST_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME])
+AML_LEGACY_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, OUTPUTS_CSV_FILENAME.format("test")])
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/utils/output_utils.py:33]: TODO: we might need to update the script that merges the outputs in cpath. I renamed the outputs files because they were all named "test_outputs.csv"

File: hi-ml-cpath/testhisto/testhisto/datamodules/test_slides_datamodule.py [hunk 6]
```diff
@@ -38,7 +39,7 @@ def mock_panda_slides_root_dir(
         src_data_path=tmp_path_to_pathmnist_dataset,
         mock_type=MockHistoDataType.PATHMNIST,
         n_tiles=1,
-        n_slides=10,
+        n_slides=16,
```
Comment (kenza-bouzid) [inline hi-ml-cpath/testhisto/testhisto/datamodules/test_slides_datamodule.py:42]: increased to be able to have more than one slide for testing/validation. The current split is .8 .1 .1

File: hi-ml-cpath/src/health_cpath/utils/output_utils.py [hunk 7]
```diff
@@ -22,15 +22,17 @@
 from health_cpath.utils.plots_utils import DeepMILPlotsHandler, TilesSelector
 from health_cpath.utils.naming import MetricsKey, ModelKey, PlotOption, ResultsKey
 
-OUTPUTS_CSV_FILENAME = "test_output.csv"
+OUTPUTS_CSV_FILENAME = "{}_output.csv"
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/output_utils.py:25]:
```suggestion
OUTPUTS_CSV_FILENAME_TEMPLATE = "{}_output.csv"
```

**Review by ant0nsc — APPROVED** [review]
some small fry comments, looks great!

File: hi-ml-cpath/src/health_cpath/datamodules/base_module.py [hunk 8]
```diff
@@ -289,8 +299,9 @@ def __init__(
         self.filter_mode = filter_mode
         # TileOnGridd transform expects None to select all foreground tile so we hardcode max_bag_size and
         # max_bag_size_inf to None if set to 0
-        self.max_bag_size = None if self.max_bag_size == 0 else self.max_bag_size  # type: ignore
-        self.max_bag_size_inf = None if self.max_bag_size_inf == 0 else self.max_bag_size_inf  # type: ignore
+        for model_key, max_bag_size in self.bag_sizes.items():
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/datamodules/base_module.py:302]: good point, I always get confused by the enum ModelKey for stages, maybe we should rename it to StageKey what do you think? (and ofc this as well)

File: hi-ml-cpath/src/health_cpath/configs/classification/BaseMIL.py [hunk 9]
```diff
@@ -26,22 +26,22 @@
...
     """BaseMIL is an abstract container defining basic functionality for running MIL experiments in both slides and
     tiles settings. It is responsible for instantiating the encoder and pooling layer. Subclasses should define the
     full DeepMIL model depending on the type of dataset (tiles/slides based).
     """
-    dropout_rate: Optional[float] = param.Number(None, bounds=(0, 1), doc="Pre-classifier dropout rate.")
     class_names: Optional[Sequence[str]] = param.List(None, item_type=str, doc="List of class names. If `None`, "
                                                                                "defaults to `('0', '1', ...)`.")
     # Data module parameters:
     batch_size: int = param.Integer(16, bounds=(1, None), doc="Number of slides to load per batch.")
+    batch_size_inf: int = param.Integer(16, bounds=(1, None), doc="Number of slides per batch during inference.")
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/configs/classification/BaseMIL.py:44]: I used the same suffix as max_bag_size_inf, can we keep it batch_size_inf for consistency?
Otherwise, will have to update it in all cpath configs/datamoules ..., seems like an unnecessary change

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py [hunk 10]
```diff
@@ -177,3 +177,25 @@ def get_pooling_layer(self, num_encoding: int) -> Tuple[nn.Module, int]:
...
         set_module_gradients_enabled(pooling_layer, tuning_flag=self.tune_pooling)
         return pooling_layer, num_features
+
+
+class ClassifierParams(param.Parameterized):
+    dropout_rate: Optional[float] = param.Number(None, bounds=(0, 1), doc="Pre-classifier dropout rate.")
+    tune_classifier: bool = param.Boolean(
+        default=True,
+        doc="If True (default), fine-tune the classifier during training. If False, keep the classifier frozen.")
+    pretrained_classifier: bool = param.Boolean(
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:187]: In general I agree that it should be named in a way that shows that it's a boolean, However I will have to rename those flags as well, and might have a lot of side effects on docs as well, I am trying to limit big changes untill I merge to main 
https://github.com/microsoft/hi-ml/blob/96141ac615beb840177b21b5a3d23c8d08ad9901/hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py#L129
https://github.com/microsoft/hi-ml/blob/96141ac615beb840177b21b5a3d23c8d08ad9901/hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py#L50

File: hi-ml-cpath/src/health_cpath/utils/output_utils.py [hunk 11]
```diff
@@ -22,15 +22,17 @@
...
+EXTRA_VAL_OUTPUTS_SUBDIR = "extra_val"
+EXTRA_PREFIX = "extra_"
 
 AML_OUTPUTS_DIR = "outputs"
-AML_LEGACY_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, OUTPUTS_CSV_FILENAME])
-AML_VAL_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, VAL_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME])
-AML_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, TEST_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME])
+AML_LEGACY_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, OUTPUTS_CSV_FILENAME.format("test")])
+AML_VAL_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, VAL_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME.format("val")])
+AML_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, TEST_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME.format("test")])
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/output_utils.py:35]: Ahh these seem to be used for HTML reports might have some side effects, I might revert

File: hi-ml-cpath/src/health_cpath/utils/output_utils.py [hunk 12]
```diff
@@ -22,15 +22,17 @@
...
 PREV_VAL_OUTPUTS_SUBDIR = "val_old"
 TEST_OUTPUTS_SUBDIR = "test"
+EXTRA_VAL_OUTPUTS_SUBDIR = "extra_val"
+EXTRA_PREFIX = "extra_"
 
 AML_OUTPUTS_DIR = "outputs"
-AML_LEGACY_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, OUTPUTS_CSV_FILENAME])
-AML_VAL_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, VAL_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME])
-AML_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, TEST_OUTPUTS_SUBDIR, OUTPUTS_CSV_FILENAME])
+AML_LEGACY_TEST_OUTPUTS_CSV = "/".join([AML_OUTPUTS_DIR, OUTPUTS_CSV_FILENAME_TEMPLATE.format("test")])
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/utils/output_utils.py:33]: @dccastro do you think we should retain from making this change, cause we won't be able to generate reports for old runs that have val/test_outputs.csv files? should we support both?

File: hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+        root_path=mock_panda_tiles_root_dir,
+        batch_size=batch_size,
+        batch_size_inf=batch_size_inf,
+        max_bag_size=16,
+        max_bag_size_inf=16,
+    )
+    _assert_correct_batch_sizes(datamodule, batch_size, batch_size_inf)
+
+
+def _assert_sampler_state(datamodule: HistoDataModule, stages: List[ModelKey], expected_none: bool) -> None:
```
Comment (dccastro) [inline hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py:116]: [Just for awareness] "Assert X" normally refers to a Boolean predicate, meaning "verify that X is true". For any other uses, something like "validate" or "check" would sound more natural.

Comment (dccastro) [reply hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py:116]: In this particular case, you could be more explicit, like `_validate_sampler_type()`.

File: hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    sampler_instances = {
```
Comment (dccastro) [inline hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py:117]: [Minor] Calling this variable `[expected_]sampler_types` might be more precise.

File: hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+from health_cpath.datamodules.panda_module import PandaSlidesDataModule, PandaTilesDataModule
+from health_cpath.utils.naming import ModelKey, SlideKey
+from health_ml.utils.common_utils import is_gpu_available
+from testhisto.utils.utils_testhisto import run_distributed
+
+
+no_gpu = not is_gpu_available()
+
+
+def _assert_correct_bag_sizes(datamodule: HistoDataModule, max_bag_size: int, max_bag_size_inf: int,
```
Comment (dccastro) [inline hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py:22]: [Minor] Shouldn't `max_bag_size_inf` be `Optional[int]`?

File: hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+                              max_expected_bag_sizes: List[int]) -> None:
```
Comment (dccastro) [inline hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py:23]: This `max_expected_bag_sizes` argument is a bit confusing. Maybe worth clarifying somehow that this refers to the full true bag sizes (assuming I understood it correctly). Both here and in the test functions that call it (e.g. with named local variables).

**Review by dccastro — APPROVED** [review]

File: hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py (hunk 16 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/testhisto/testhisto/datamodules/test_histo_datamodule.py:23]: Good point. wiil do!
